Constraint distances mangled after `Interchange.combine`

(More info coming later)

**Description**

Please describe what you were trying to do, what you expected to happen, and what happened instead.

**Reproduction**

Please include a [minimally reproducing example](https://stackoverflow.com/help/minimal-reproducible-example) of this bug.
```python
In [1]: from openff.toolkit import ForceField, Molecule

In [2]: int1 = ForceField("openff-2.2.1.offxml").create_interchange(Molecule.fro
      ⋮ m_smiles("CCO").to_topology())
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/amber_ff_ports/amber_ff_ports.py:8: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import resource_filename
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/amber_ff_ports/amber_ff_ports.py:8: UserWarning: Module openff was already imported from None, but /Users/mattthompson/software/openff-nagl-models is being added to sys.path
  from pkg_resources import resource_filename

In [3]: int2 = ForceField("openff-1.0.0.offxml").create_interchange(Molecule.fro
      ⋮ m_smiles("CCO").to_topology())
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/smirnoff/_valence.py:48: ForceFieldModificationWarning: Automatically up-converting BondHandler from version 0.3 to 0.4. Consider manually upgrading this BondHandler (or <Bonds> section in an OFFXML file) to 0.4 or newer. For more details, see https://openforcefield.github.io/standards/standards/smirnoff/#bonds.
  warnings.warn(

In [4]: int1['Constraints'].get_system_parameters()
Out[4]:
array([[1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [0.97537481]])

In [5]: int2['Constraints'].get_system_parameters()
Out[5]:
array([[1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [0.97076876]])

In [6]: int1.combine(int2)['Constraints'].get_system_parameters()
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/operations/_combine.py:100: InterchangeCombinationWarning: Interchange object combination is complex and may produce strange results outside of use cases it has been tested in. Use with caution and thoroughly validate results!
  warnings.warn(
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/operations/_combine.py:80: InterchangeCombinationWarning: Found electrostatics 1-4 scaling factors of 5/6 with slightly different rounding (0.833333 and 0.8333333333). This likely stems from OpenFF using more digits in rounding 1/1.2. The value of 0.8333333333 will be used, which may or may not introduce small errors.
  warnings.warn(
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/operations/_combine.py:170: InterchangeCombinationWarning: Setting positions to None because one or both objects added together were missing positions.
  warnings.warn(
Out[6]:
array([[1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [0.97076876],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [1.09288838],
       [0.97076876]])

In [7]: int2.combine(int1)['Constraints'].get_system_parameters()
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/operations/_combine.py:100: InterchangeCombinationWarning: Interchange object combination is complex and may produce strange results outside of use cases it has been tested in. Use with caution and thoroughly validate results!
  warnings.warn(
/Users/mattthompson/.local/share/mamba/envs/openff-toolkit-test-rdkit/lib/python3.12/site-packages/openff/interchange/operations/_combine.py:170: InterchangeCombinationWarning: Setting positions to None because one or both objects added together were missing positions.
  warnings.warn(
Out[7]:
array([[1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [0.97537481],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [1.09397889],
       [0.97537481]])
```

**Output**

Please include the output, including full tracebacks of any errors, resulting from the reproduction.

**Software versions**

- Which operating system and version are you using?
- How did you install Interchange?
- What is the output of running `conda list`?

The problem is a key collision which is not checked for. In this case, the problem arises when two SMIRKS patterns from different force fields refer to different parameters, in this case bond/constraint distances.

Applying this diff to the reproduction above

```diff
diff --git a/openff/interchange/operations/_combine.py b/openff/interchange/operations/_combine.py
index 0fb227e9..23e708da 100644
--- a/openff/interchange/operations/_combine.py
+++ b/openff/interchange/operations/_combine.py
@@ -147,6 +147,14 @@ def _combine(
                     _tmp_pot_key.mult = _mult
                     _mult += 1
 
+            assert new_top_key not in self_collection.key_map, (
+                f"Trying to add a topology key {new_top_key} which is already present in this collection {self_collection=}"
+            )
+
+            assert pot_key not in self_collection.potentials, (
+                f"Trying to add a potential key {pot_key} which is already present in this collection {self_collection=}"
+            )
+
             self_collection.key_map.update({new_top_key: _tmp_pot_key})
             if collection_name == "Constraints":
                 self_collection.potentials.update(
```

Throws

```
AssertionError: Trying to add a potential key id='[#6X4:1]-[#6X4:2]' mult=None associated_handler='Bonds' bond_order=None virtual_site_type=None cosmetic_attributes={} which is already present in this collection self_collection=Handler 'Bonds' with expression 'k/2*(r-length)**2', 8 mapping keys, and 4 potentials

--- a/docs/releasehistory.md
+++ b/docs/releasehistory.md
@@ -11,6 +11,19 @@ Dates are given in YYYY-MM-DD format.
 
 Please note that all releases prior to a version 1.0.0 are considered pre-releases and many API changes will come before a stable release.
 
+## 0.4.6 - 2025-09-02
+
+### Bug fixes
+
+* #1312 Fix handling of `DoubleExponentialVirtualSites`
+* #1315 Fixes a bug in which `Interchange.combine` did not properly process constraint distances.
+
+### Maintenance
+
+* #1297 Do not duplicate type annotations in docstrings
+* #1304 Automatically format `pyproject.toml`
+* #1306 Declare minimum Python version in project metadata
+
 ## 0.4.5 - 2025-08-20
 
 ### New features

--- a/openff/interchange/_tests/unit_tests/operations/test_combine.py
+++ b/openff/interchange/_tests/unit_tests/operations/test_combine.py
@@ -192,8 +192,9 @@ class TestCombine:
             )
 
     @pytest.mark.parametrize("flip_order", [False, True])
-    def test_constraint_key_collision(self, parsley, sage, ethanol, flip_order):
-        """Test that key collisions in constraints are handled."""
+    @pytest.mark.parametrize("handler_name", ["Constraints", "Bonds", "Angles", "ProperTorsions"])
+    def test_constraint_key_collision(self, parsley, sage, ethanol, flip_order, handler_name):
+        """Test that key collisions in constraints and valence terms are handled."""
         interchanges = [
             parsley.create_interchange(
                 ethanol.to_topology(),
@@ -207,12 +208,14 @@ class TestCombine:
         if flip_order:
             interchanges.reverse()
 
-        constraint_arrays = [interchange["Constraints"].get_system_parameters() for interchange in interchanges]
+        arrays_before = [interchange[handler_name].get_system_parameters() for interchange in interchanges]
 
-        array_after_combine = interchanges[0].combine(interchanges[1])["Constraints"].get_system_parameters()
+        numpy.testing.assert_raises(AssertionError, numpy.testing.assert_allclose, *arrays_before)
+
+        array_after_combine = interchanges[0].combine(interchanges[1])[handler_name].get_system_parameters()
 
         # check that the contents of the combined Interchange contains each input, without mushing
         numpy.testing.assert_allclose(
-            numpy.asarray(constraint_arrays).reshape((12, 1)),
+            numpy.vstack(arrays_before),
             array_after_combine,
         )

--- a/openff/interchange/_tests/unit_tests/operations/test_combine.py
+++ b/openff/interchange/_tests/unit_tests/operations/test_combine.py
@@ -190,3 +190,29 @@ class TestCombine:
             ).combine(
                 sage.create_interchange(ethanol.to_topology()),
             )
+
+    @pytest.mark.parametrize("flip_order", [False, True])
+    def test_constraint_key_collision(self, parsley, sage, ethanol, flip_order):
+        """Test that key collisions in constraints are handled."""
+        interchanges = [
+            parsley.create_interchange(
+                ethanol.to_topology(),
+            ),
+            sage.create_interchange(
+                ethanol.to_topology(),
+            ),
+        ]
+
+        # want to make sure this behavior is not order-dependent
+        if flip_order:
+            interchanges.reverse()
+
+        constraint_arrays = [interchange["Constraints"].get_system_parameters() for interchange in interchanges]
+
+        array_after_combine = interchanges[0].combine(interchanges[1])["Constraints"].get_system_parameters()
+
+        # check that the contents of the combined Interchange contains each input, without mushing
+        numpy.testing.assert_allclose(
+            numpy.asarray(constraint_arrays).reshape((12, 1)),
+            array_after_combine,
+        )

--- a/openff/interchange/operations/_combine.py
+++ b/openff/interchange/operations/_combine.py
@@ -151,11 +151,6 @@ def _combine(
                     _tmp_pot_key.mult = _mult
                     _mult += 1
 
-            if new_top_key in self_collection.key_map:
-                raise UnsupportedCombinationError(
-                    "Key collision detected, please raise an issue",
-                )
-
             if _tmp_pot_key in self_collection.potentials:
                 logging.info(f"Key collision, fixing. Key is {_tmp_pot_key}")
                 _tmp_pot_key.id = _tmp_pot_key.id + "_DUPLICATE"

--- a/openff/interchange/operations/_combine.py
+++ b/openff/interchange/operations/_combine.py
@@ -1,6 +1,7 @@
 """The logic behind `Interchange.combine`."""
 
 import copy
+import logging
 import warnings
 from typing import TYPE_CHECKING
 
@@ -17,6 +18,9 @@ from openff.interchange.warnings import InterchangeCombinationWarning
 
 if TYPE_CHECKING:
     from openff.interchange.components.interchange import Interchange
+
+
+logger = logging.getLogger(__name__)
 
 DEFAULT_CUTOFF_TOLERANCE = Quantity("1e-6 nanometer")
 
@@ -146,6 +150,15 @@ def _combine(
                 while _tmp_pot_key in self_collection.potentials:
                     _tmp_pot_key.mult = _mult
                     _mult += 1
+
+            if new_top_key in self_collection.key_map:
+                raise UnsupportedCombinationError(
+                    "Key collision detected, please raise an issue",
+                )
+
+            if _tmp_pot_key in self_collection.potentials:
+                logging.info(f"Key collision, fixing. Key is {_tmp_pot_key}")
+                _tmp_pot_key.id = _tmp_pot_key.id + "_DUPLICATE"
 
             self_collection.key_map.update({new_top_key: _tmp_pot_key})
             if collection_name == "Constraints":

